File: FlutterProjects/Practice/Different Style/hr_and_payroll_management - v2/lib/features/attendance/AttendanceService.dart
import 'dart:convert';
import 'package:hr_and_payroll_management/features/attendance/AttendanceModel.dart';
import 'package:http/http.dart' as http;
import '../user/model/User.dart';

class AttendanceService {
  final String baseUrl = 'http://localhost:8080/api/attendance';

  Future<List<Attendance>> getAllAttendances() async {
    final response = await http.get(Uri.parse('$baseUrl/'));

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load attendances');
    }
  }

  Future<Attendance> checkIn(int userId) async {
    final response = await http.post(
      Uri.parse('$baseUrl/checkin'),
      body: json.encode({'userId': userId}),
      headers: {'Content-Type': 'application/json'},
    );

    if (response.statusCode == 201) {
      return Attendance.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to check in');
    }
  }

  Future<Attendance> checkOut(int userId) async {
    final response = await http.put(
      Uri.parse('$baseUrl/checkout'),
      body: json.encode({'userId': userId}),
      headers: {'Content-Type': 'application/json'},
    );

    if (response.statusCode == 200) {
      return Attendance.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to check out');
    }
  }

  Future<List<Attendance>> getOvertimeByUser(int userId, DateTime startDate, DateTime endDate) async {
    final response = await http.get(
      Uri.parse('$baseUrl/overtime/$userId?startDate=${startDate.toIso8601String()}&endDate=${endDate.toIso8601String()}'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load overtime records');
    }
  }

  Future<List<Attendance>> getTodayAttendance() async {
    final response = await http.get(Uri.parse('$baseUrl/today'));

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load today’s attendance');
    }
  }

  Future<Attendance> findAttendanceById(int id) async {
    final response = await http.get(Uri.parse('$baseUrl/find/$id'));

    if (response.statusCode == 200) {
      return Attendance.fromJson(json.decode(response.body));
    } else {
      throw Exception('Attendance not found');
    }
  }

  Future<List<Attendance>> getAttendancesByUserId(int userId) async {
    final response = await http.get(Uri.parse('$baseUrl/user/$userId/attendances'));

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load user attendance records');
    }
  }

  Future<Map<String, dynamic>> getUsersWithAttendanceInRange(DateTime startDate, DateTime endDate) async {
    final response = await http.get(
      Uri.parse('$baseUrl/attendanceRange?startDate=${startDate.toIso8601String()}&endDate=${endDate.toIso8601String()}'),
    );

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load attendance in range');
    }
  }

  Future<List<User>> getUsersWithoutAttendanceToday() async {
    final response = await http.get(Uri.parse('$baseUrl/usersWithoutAttendanceToday'));

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => User.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load users without attendance today');
    }
  }

  Future<List<Object>> getPeakAttendanceDay() async {
    final response = await http.get(Uri.parse('$baseUrl/peakAttendanceDay'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load peak attendance day');
    }
  }

  Future<List<Object>> getPeakAttendanceMonth() async {
    final response = await http.get(Uri.parse('$baseUrl/peakAttendanceMonth'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load peak attendance month');
    }
  }

  Future<List<Attendance>> getLateCheckIns(String lateTime, DateTime startDate, DateTime endDate) async {
    final response = await http.get(
      Uri.parse('$baseUrl/lateCheckIns?lateTime=$lateTime&startDate=${startDate.toIso8601String()}&endDate=${endDate.toIso8601String()}'),
    );

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load late check-ins');
    }
  }

  Future<List<Attendance>> getAttendancesByUserNamePart(String name) async {
    final response = await http.get(Uri.parse('$baseUrl/searchByName?name=$name'));

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load attendances by user name');
    }
  }

  Future<List<Attendance>> getAttendanceByRole(String role) async {
    final response = await http.get(Uri.parse('$baseUrl/roleAttendance?role=$role'));

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load attendances by role');
    }
  }

  Future<List<Attendance>> getTodayAttendanceByUserId(int userId) async {
    final response = await http.get(Uri.parse('$baseUrl/todayAttendance/$userId'));

    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Attendance.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load today’s attendance by user');
    }
  }
}

File: FlutterProjects/Practice/API Backup/Backend with JWT- v1 implement/src/main/java/com/mrahmed/HRandPayrollManagementSystem/service/AttendanceService.java
package com.mrahmed.HRandPayrollManagementSystem.service;

import com.mrahmed.HRandPayrollManagementSystem.entity.Attendance;
import com.mrahmed.HRandPayrollManagementSystem.entity.User;
import com.mrahmed.HRandPayrollManagementSystem.repository.AttendanceRepository;
import com.mrahmed.HRandPayrollManagementSystem.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class AttendanceService {

    @Autowired
    private  AttendanceRepository attendanceRepository;

    @Autowired
    private  UserRepository userRepository;

    public List<Attendance> getOvertimeForUser(Long userId, LocalDate startDate, LocalDate endDate) {
        List<Attendance> attendances = attendanceRepository.findAttendancesByUserIdAndDateRange(userId, startDate, endDate);
        return attendances.stream()
                .filter(this::isOvertime)
                .collect(Collectors.toList());
    }

    // Check if the attendance record contains overtime
    private boolean isOvertime(Attendance attendance) {
        if (attendance.getClockInTime() == null || attendance.getClockOutTime() == null) {
            return false; // Skip records with missing times
        }
        Duration duration = Duration.between(attendance.getClockInTime(), attendance.getClockOutTime());
        long hours = duration.toHours();
        return hours > 8; // Assuming a workday is 8 hours
    }

    public Attendance checkIn(long userId) {
        try {
            User user = userRepository.findById(userId)
                    .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));
            if (attendanceRepository.countByUserIdAndDate(userId, LocalDate.now()) > 0) {
                throw new RuntimeException("User has already checked in today.");
            }
            Attendance attendance = new Attendance();
            attendance.setDate(LocalDate.now());
            attendance.setClockInTime(LocalDateTime.now());
            attendance.setUser(user);
            return attendanceRepository.save(attendance);
        } catch (Exception e) {
            throw new RuntimeException("Error occurred while checking in: " + e.getMessage(), e);
        }
    }

    public Attendance checkOut(long userId) {
        try {
            Attendance attendance = attendanceRepository.findLastAttendanceForUser(userId, LocalDate.now())
                    .orElseThrow(() -> new RuntimeException("No check-in found for today's check-out"));
            if (attendance.getClockOutTime() != null) {
                throw new RuntimeException("User has already checked out today.");
            }
            attendance.setClockOutTime(LocalDateTime.now());
            return attendanceRepository.save(attendance);
        } catch (Exception e) {
            throw new RuntimeException("Error occurred while checking out: " + e.getMessage(), e);
        }
    }


    // Fetch today's attendance
    public List<Attendance> getTodayAttendances() {
        LocalDate today = LocalDate.now(); // Get the current date
        return attendanceRepository.findAttendancesForToday(today); // Fetch today's attendance
    }

    // Fetch all attendance records
    public List<Attendance> getAllAttendances() {
        return attendanceRepository.findAll();
    }

    // Find attendance by ID
    public Attendance findAttendanceById(long id) {
        return attendanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Attendance not found with id: " + id));
    }

    public Map<User, Long> getUsersAttendanceInRange(LocalDate startDate, LocalDate endDate) {
        List<Attendance> attendances = attendanceRepository.findAttendancesInRange(startDate, endDate);
        return attendances.stream()
                .collect(Collectors.groupingBy(Attendance::getUser, Collectors.counting()));
    }


    // Get all attendance records for a specific user
    public List<Attendance> getAttendanceByUserId(Long id) {
        return attendanceRepository.findAllByUserId(id);
    }

    // Find users without attendance today
    public List<User> findUsersWithoutAttendanceToday() {
        return attendanceRepository.findUsersWithoutAttendanceForToday(LocalDate.now());
    }

    // Fetch peak attendance day, month, and year
    public List<Object[]> getPeakAttendanceDay() {
        return attendanceRepository.findPeakAttendanceDay();
    }

    public List<Object[]> getPeakAttendanceMonth() {
        return attendanceRepository.findPeakAttendanceMonth();
    }


    // Get late check-ins
    public List<Attendance> getLateCheckIns(String lateTime, LocalDate startDate, LocalDate endDate) {
        return attendanceRepository.findLateCheckIns(lateTime, startDate, endDate);
    }


    // Search attendance records by userName
    public List<Attendance> getAttendancesByUserNamePart(String name) {
        return attendanceRepository.findAttendancesByUserNamePart(name);
    }

    // Filter attendance by user role and date range
    public List<Attendance> getAttendanceByRole(String role) {
        return attendanceRepository.findAttendanceByRole(role);
    }

    // Get today's attendance by user ID
    public List<Attendance> getTodayAttendanceByUserId(long userId) {
        return attendanceRepository.findTodayAttendanceByUserId(userId);
    }

}